# Exercise: Imbalanced data model bias

In this exercise, we will take a closer look at *imbalanced datasets*, what effects they have on predictions and how they can be addressed.

We will also employ *confusion matrices* to evaluate model updates.

## Data visualization

Just like in the previous exercise, we use a dataset that represents different classes of objects found on the mountain:

In [1]:
import pandas

#Import the data from the .csv file
dataset = pandas.read_csv('snow_objects.csv', delimiter="\t")

# Let's have a look at the data
dataset

size  roughness  color    motion   label
0     50.959361   1.318226  green  0.054290    tree
1     60.008521   0.554291  brown  0.000000    tree
2     20.530772   1.097752  white  1.380464    tree
3     28.092138   0.966482   grey  0.650528    tree
4     48.344211   0.799093   grey  0.000000    tree
...         ...        ...    ...       ...     ...
2195   1.918175   1.182234  white  0.000000  animal
2196   1.000694   1.332152  black  4.041097  animal
2197   2.331485   0.734561  brown  0.961486  animal
2198   1.786560   0.707935  black  0.000000  animal
2199   1.518813   1.447957  brown  0.000000  animal

[2200 rows x 5 columns]

Recall that we have an *imbalanced dataset*. Some classes are much more frequent than others:

In [2]:
import graphings

# Plot a histogram with counts for each label
graphings.multiple_histogram(dataset, label_x="label", label_group="label", title="Label distribution")


## Using binary classification

For this exercise we will build a *binary classification model*. We want to predict if objects in the snow are "hikers" or "not-hikers".

To do that, we first need to add another column to our dataset, and set it to `True` where the original label is `hiker`, and `False` to anything else:


In [3]:
# Add a new label with true/false values to our dataset
dataset["is_hiker"]= dataset.label == "hiker"
dataset.head()

size  roughness  color    motion label  is_hiker
0  50.959361   1.318226  green  0.054290  tree     False
1  60.008521   0.554291  brown  0.000000  tree     False
2  20.530772   1.097752  white  1.380464  tree     False
3  28.092138   0.966482   grey  0.650528  tree     False
4  48.344211   0.799093   grey  0.000000  tree     False

In [4]:
# Plot frequency for new label
graphings.multiple_histogram(dataset, label_x="is_hiker", label_group="is_hiker", title="Distribution for new binary label 'is_hiker'")

We now have only two classes of labels in our dataset, but we have made it even more imbalanced.

Let's train the random forest model using `is_hiker` as the target variable, then measure its accuracy on both *train* and *test* sets:

In [5]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
# import matplotlib.pyplot as plt
# from sklearn.metrics import plot_confusion_matrix
from sklearn.metrics import accuracy_score

# Custom function that measures accuracy on different models and datasets
# We will use this in different parts of the exercise
def assess_accuracy(model, dataset, label):
    """
    Asesses model accuracy on different sets
    """

    actual = dataset[label]
    predictions = model.predict(dataset[features])
    acc= accuracy_score(actual, predictions)
    return acc

# Split the dataset in an 70/30 train/test ratio. 
train, test = train_test_split(dataset, test_size=0.3, random_state=1, shuffle=True)

# define a random forest model
model = RandomForestClassifier(n_estimators=1, random_state=1, verbose=False)

# Define which features are to be used (leave color out for now)
features = ["size", "roughness", "motion"]

# Train the model using the binary label
model.fit(train[features], train.is_hiker)

print("Train accuracy:", assess_accuracy(model,train, "is_hiker"))
print("Test accuracy:", assess_accuracy(model,test, "is_hiker"))

Train accuracy: 0.9532467532467532
Test accuracy: 0.906060606060606


Accuracy looks good for both *train* and *test* sets, but remember that this metric is not an absolute measure of success.

We should plot a confusion matrix to see how the model is actually doing:

In [6]:
# sklearn has a very convenient utility to build confusion matrices
from sklearn.metrics import confusion_matrix
import plotly.figure_factory as ff

# Calculate the model's accuracy on the TEST set
actual = test.is_hiker
predictions = model.predict(test[features])

# Build and print our confusion matrix, using the actual values and predictions 
# from the test set, calculated in previous cells
cm = confusion_matrix(actual, predictions, normalize=None)

# Create the list of unique labels in the test set, to use in our plot
# I.e., ['True', 'False',]
unique_targets = sorted(list(test["is_hiker"].unique()))

# Convert values to lower case so the plot code can count the outcomes
x=y= [str(s).lower() for s in unique_targets]
print(x)
print(y)

['false', 'true']
['false', 'true']


In [7]:
# Plot the matrix above as a heatmap with annotations (values) in its cells
fig = ff.create_annotated_heatmap(cm, x, y)

# Set titles and ordering
fig.update_layout(  title_text="<b>Confusion matrix</b>", 
                    yaxis = dict(categoryorder = "category descending")
                    )

fig.add_annotation(dict(font=dict(color="black",size=14),
                        x=0.5,
                        y=-0.15,
                        showarrow=False,
                        text="Predicted label",
                        xref="paper",
                        yref="paper"))

fig.add_annotation(dict(font=dict(color="black",size=14),
                        x=-0.15,
                        y=0.5,
                        showarrow=False,
                        text="Actual label",
                        textangle=-90,
                        xref="paper",
                        yref="paper"))

# We need margins so the titles fit
fig.update_layout(margin=dict(t=80, r=20, l=120, b=50))
fig['data'][0]['showscale'] = True
fig.show()

The confusion matrix shows us that, despite the reported metrics, the model is not incredibly precise.

Out of the 660 samples present in the *test* set (30% of the total samples), it predicted `29` *false negatives* and `33` _false positives_.

More importantly, look at the bottom row, which shows what happened when the model was shown information about a hiker: it got the answer wrong almost 30% of the time. This means it would not correctly identify almost 30% of the people on the mountain!  

What happens if we used this model to make predictions on balanced sets?

Let's load a dataset with an equal number of outcomes for "hikers" and "non-hikers", then use that data to make predictions:

In [8]:
# Load and print umbiased set
#Import the data from the .csv file
balanced_dataset = pandas.read_csv('snow_objects_balanced.csv', delimiter="\t")

#Let's have a look at the data
graphings.multiple_histogram(balanced_dataset, label_x="label", label_group="label", title="Label distribution")

This new dataset is balanced among the classes, but for our purposes we want it balanced between hikers and non-hikers. 

For simplicity, let's take the hikers plus a random sampling of the non-hikers.

In [9]:
# Add a new label with true/false values to our dataset
balanced_dataset["is_hiker"]= balanced_dataset.label == "hiker"
balanced_dataset["is_hiker"].head()

0    False
1    False
2    False
3    False
4    False
Name: is_hiker, dtype: bool

In [13]:
balanced_dataset["is_hiker"]== 1

0       False
1       False
2       False
3       False
4       False
        ...  
1995    False
1996    False
1997    False
1998    False
1999    False
Name: is_hiker, Length: 2000, dtype: bool

In [12]:
balanced_dataset["is_hiker"]== 0

0       True
1       True
2       True
3       True
4       True
        ... 
1995    True
1996    True
1997    True
1998    True
1999    True
Name: is_hiker, Length: 2000, dtype: bool

In [14]:
balanced_dataset[balanced_dataset["is_hiker"]== 1]

size  roughness   color    motion  label  is_hiker
1000  1.758147   0.578384   green  1.979197  hiker      True
1001  2.138218   1.096874  orange  2.103307  hiker      True
1002  1.729903   1.213756   brown  2.139579  hiker      True
1003  1.614283   0.889785   green  1.402956  hiker      True
1004  2.202475   0.933503   brown  2.337197  hiker      True
...        ...        ...     ...       ...    ...       ...
1495  1.381068   0.926941   brown  3.644156  hiker      True
1496  1.313774   0.783504   brown  1.412850  hiker      True
1497  1.595399   0.628819   brown  0.000000  hiker      True
1498  1.809689   0.192637   green  4.262440  hiker      True
1499  2.092057   0.809787   black  2.952339  hiker      True

[500 rows x 6 columns]

In [15]:
balanced_dataset[balanced_dataset["is_hiker"]== 0]

size  roughness  color    motion   label  is_hiker
0     61.462366   0.901306  brown  0.000000    tree     False
1     56.244378   0.380033  brown  0.000000    tree     False
2     31.407699   0.929223  brown  0.570701    tree     False
3     26.629504   0.320678  green  0.833444    tree     False
4     68.340183   1.262819  brown  0.000000    tree     False
...         ...        ...    ...       ...     ...       ...
1995   1.960167   0.932040  brown  6.841839  animal     False
1996   1.507176   0.585017  black  7.962275  animal     False
1997   1.348828   0.900110  white  5.334002  animal     False
1998   1.944979   0.856855  black  0.000000  animal     False
1999   1.494498   0.000000  white  4.878107  animal     False

[1500 rows x 6 columns]

In [16]:

hikers_dataset = balanced_dataset[balanced_dataset["is_hiker"]== 1]
nonhikers_dataset = balanced_dataset[balanced_dataset["is_hiker"]==False]
# take a random sampling of non-hikers the same size as the hikers subset
len(hikers_dataset.index)

500

In [17]:
nonhikers_dataset = nonhikers_dataset.sample(n=len(hikers_dataset.index), random_state=1)
balanced_dataset = pandas.concat([hikers_dataset, nonhikers_dataset])
balanced_dataset

size  roughness   color    motion   label  is_hiker
1000   1.758147   0.578384   green  1.979197   hiker      True
1001   2.138218   1.096874  orange  2.103307   hiker      True
1002   1.729903   1.213756   brown  2.139579   hiker      True
1003   1.614283   0.889785   green  1.402956   hiker      True
1004   2.202475   0.933503   brown  2.337197   hiker      True
...         ...        ...     ...       ...     ...       ...
1654   1.003537   1.287778   white  0.214379  animal     False
1544   1.543077   1.186846   black  2.723995  animal     False
323   31.716681   1.384797   brown  0.337607    tree     False
534    0.495596   1.384001   brown  0.013193    rock     False
1624   1.868036   0.886488   black  5.006563  animal     False

[1000 rows x 6 columns]

In [18]:
# Plot frequency for "is_hiker" labels
graphings.multiple_histogram(balanced_dataset, label_x="is_hiker", label_group="is_hiker", title="Label distribution in balanced dataset")

As you can see, the `is_hiker` label has the same number of `True` and `False` for both classes. We are now using a *class balanced dataset*.

Let's run predictions on this set using the previously trained model:

In [19]:
# Test the model using a balanced dataset
actual = balanced_dataset.is_hiker
predictions = model.predict(balanced_dataset[features])

# Build and print our confusion matrix, using the actual values and predictions 
# from the test set, calculated in previous cells
cm = confusion_matrix(actual, predictions, normalize=None)

# Print accuracy using this set
print("Balanced set accuracy:", assess_accuracy(model,balanced_dataset, "is_hiker"))

Balanced set accuracy: 0.754


As expected, we see a noticeable drop in accuracy using a different set.

Again, let's visually analyze its performance:

In [20]:
# plot new confusion matrix
# Create the list of unique labels in the test set to use in our plot
unique_targets = sorted(list(balanced_dataset["is_hiker"].unique()))

# Convert values to lower case so the plot code can count the outcomes
x = y = [str(s).lower() for s in unique_targets]

# Plot the matrix above as a heatmap with annotations (values) in its cells
fig = ff.create_annotated_heatmap(cm, x, y)

# Set titles and ordering
fig.update_layout(  title_text="<b>Confusion matrix</b>", 
                    yaxis = dict(categoryorder = "category descending")
                    )

fig.add_annotation(dict(font=dict(color="black",size=14),
                        x=0.5,
                        y=-0.15,
                        showarrow=False,
                        text="Predicted label",
                        xref="paper",
                        yref="paper"))

fig.add_annotation(dict(font=dict(color="black",size=14),
                        x=-0.15,
                        y=0.5,
                        showarrow=False,
                        text="Actual label",
                        textangle=-90,
                        xref="paper",
                        yref="paper"))

# We need margins so the titles fit
fig.update_layout(margin=dict(t=80, r=20, l=120, b=50))
fig['data'][0]['showscale'] = True
fig.show()

The confusion matrix confirms the poor accuracy using this dataset, but why is this happening when we had such excellent metrics in the earlier *train* and *test* sets?

Recall that the first model was heavily imbalanced. The "hiker" class made up roughly 22% of the outcomes.

When such an imbalance happens, classification models don't have enough data to learn the patterns for the minority __class__, and as a consequence become biased towards the __majority__ class!

Imbalanced sets can be addressed in a number of ways:

- Improving data selection
- Resampling the dataset
- Using weighted classes

For this exercise, we will focus on the last option.

## Using class weights to balance dataset

We can assign different *weights* to the majority and minority classes, according to their distribution, and modify our training algorithm so that it takes that information into account during the training phase.

It will then penalize errors when the minority class is misclassified, in essence "forcing" the model to to better learn their features and patterns.

To use weighted classes, we have to retrain our model using the original *train* set, but this time telling the algorithm to use weights when calculating errors:


In [21]:
# Import function used in calculating  weights
from sklearn.utils import class_weight

# Retrain model using class weights
# Using class_weight="balanced" tells the algorithm to automatically calculate weights for us
weighted_model = RandomForestClassifier(n_estimators=1, random_state=1, verbose=False, class_weight="balanced")
# Train the weighted_model using binary label
weighted_model.fit(train[features], train.is_hiker)

print("Train accuracy:", assess_accuracy(weighted_model,train, "is_hiker"))
print("Test accuracy:", assess_accuracy(weighted_model, test, "is_hiker"))

Train accuracy: 0.9525974025974026
Test accuracy: 0.9166666666666666


After using the weighted classes, the *train* accuracy remained almost the same, while the *test* accuracy showed a small improvement (roughly 1%).

Let's see if results are improved at all using the __balanced__ set for predictions again:


In [22]:
print("Balanced set accuracy:", assess_accuracy(weighted_model, balanced_dataset, "is_hiker"))

# Test the weighted_model using a balanced dataset
actual = balanced_dataset.is_hiker
predictions = weighted_model.predict(balanced_dataset[features])

# Build and print our confusion matrix, using the actual values and predictions 
# from the test set, calculated in previous cells
cm = confusion_matrix(actual, predictions, normalize=None)


Balanced set accuracy: 0.796


The accuracy for the balanced set increased roughly 4%, but we should still try to to visualize and understand the new results.

## Final confusion matrix

We can now plot a final confusion matrix, representing predictions for a *balanced dataset*, using a model trained on a *weighted class dataset*:

In [23]:
# Plot the matrix above as a heatmap with annotations (values) in its cells
fig = ff.create_annotated_heatmap(cm, x, y)

# Set titles and ordering
fig.update_layout(  title_text="<b>Confusion matrix</b>", 
                    yaxis = dict(categoryorder = "category descending")
                    )

fig.add_annotation(dict(font=dict(color="black",size=14),
                        x=0.5,
                        y=-0.15,
                        showarrow=False,
                        text="Predicted label",
                        xref="paper",
                        yref="paper"))

fig.add_annotation(dict(font=dict(color="black",size=14),
                        x=-0.15,
                        y=0.5,
                        showarrow=False,
                        text="Actual label",
                        textangle=-90,
                        xref="paper",
                        yref="paper"))

# We need margins so the titles fit
fig.update_layout(margin=dict(t=80, r=20, l=120, b=50))
fig['data'][0]['showscale'] = True
fig.show()

While the results might look a bit disappointing, we now have 21% wrong predictions (FNs + FPs), against 25% from the previous experiment.

Correct predictions (TPs + TNs) went from 74.7% to 78.7%.

Is an all around 4% improvement significant or not?

Remember that we had relatively little data to train the model, and the features we have available may still be so similar for different samples (for example, hikers and animals tend to be small, non-rough and move a lot), that despite our efforts, the model still has some difficulty making correct predictions.

We only had to change a single line of code to get better results, so it seems worth the effort!


## Summary

This was a long exercise, where we covered the following topics:

- Creating new label fields so we can perform *binary classification* using a dataset with multiple classes.
- How training on *imbalanced sets* can have a negative effect in perfomance, especially when using unseen data from *balanced datasets*.
- Evaluating results of *binary classification* models using a confusion matrix.
- Using weighted classes to address class imbalances when training a model and evaluating the results.